Add run / model links to MLP train step card

dbczumar opened the pull request:
<!-- 🚨 We recommend pull requests be filed from a non-master branch on a repository fork (e.g. <username>:fix-xxx). 🚨 -->

## What changes are proposed in this pull request?

Add run / model links to MLP train step card on Databricks

## How is this patch tested?

- Unit tests
- Manually verified that card links appear on Databricks, preexisting text appears outside of Databricks

<img width="1407" alt="Screen Shot 2022-07-26 at 12 32 25 PM" src="https://user-images.githubusercontent.com/39497902/181098804-bec685f3-29b0-4a4d-bae6-7d918fbc15b7.png">

<img width="911" alt="Screen Shot 2022-07-26 at 12 32 21 PM" src="https://user-images.githubusercontent.com/39497902/181098853-f855b75a-2613-4f92-829c-1cd37613b1ca.png">

<img width="699" alt="Screen Shot 2022-07-26 at 12 32 05 PM" src="https://user-images.githubusercontent.com/39497902/181098926-76dbcb4a-da94-4ae0-af97-aaf8e4406c8f.png">

<img width="2301" alt="Screen Shot 2022-07-26 at 12 39 21 PM" src="https://user-images.githubusercontent.com/39497902/181098979-9f375efa-e2da-464d-b1a9-63a848572f5f.png">

<img width="1775" alt="Screen Shot 2022-07-26 at 12 39 16 PM" src="https://user-images.githubusercontent.com/39497902/181098938-453b0f06-7cf1-4047-a4b1-73be3ede0542.png">

<img width="1221" alt="Screen Shot 2022-07-26 at 12 39 27 PM" src="https://user-images.githubusercontent.com/39497902/181099005-0aac53b9-f6f4-44a8-b5a2-240fb8a80506.png">

- Verified that model registration source link remains correct:

Created "remote workspace" (still E2 Dogfood) secrets

```
$ databricks secrets create-scope --scope foo
$ databricks secrets put --scope foo --key bar-host
$ databricks secrets put --scope foo --key bar-token
$ databricks secrets put --scope foo --key bar-workspace-id
```

Registered a model using a run from the "remote workspace" and verified that the resulting run link is as expected:

<img width="1977" alt="Screen Shot 2022-07-26 at 1 15 50 PM" src="https://user-images.githubusercontent.com/39497902/181105365-155a29a7-5d80-44fe-8487-5dfa452a49a6.png">

## Does this PR change the documentation?

- [X] No. You can skip the rest of this section.
- [ ] Yes. Make sure the changed pages / sections render correctly by following the steps below.

1. Check the status of the `ci/circleci: build_doc` check. If it's successful, proceed to the
   next step, otherwise fix it.
2. Click `Details` on the right to open the job page of CircleCI.
3. Click the `Artifacts` tab.
4. Click `docs/build/html/index.html`.
5. Find the changed pages / sections and make sure they render correctly.

## Release Notes

Add run / model links to the train step card on Databricks.

### Is this a user-facing change?

- [ ] No. You can skip the rest of this section.
- [X] Yes. Give a description of this change to be included in the release notes for MLflow users.

(Details in 1-2 sentences. You can just refer to another PR with a description if this PR is part of a larger change.)

### What component(s), interfaces, languages, and integrations does this PR affect?
Components
- [ ] `area/artifacts`: Artifact stores and artifact logging
- [ ] `area/build`: Build and test infrastructure for MLflow
- [ ] `area/docs`: MLflow documentation pages
- [ ] `area/examples`: Example code
- [ ] `area/model-registry`: Model Registry service, APIs, and the fluent client calls for Model Registry
- [ ] `area/models`: MLmodel format, model serialization/deserialization, flavors
- [x] `area/pipelines`: Pipelines, Pipeline APIs, Pipeline configs, Pipeline Templates
- [ ] `area/projects`: MLproject format, project running backends
- [ ] `area/scoring`: MLflow Model server, model deployment tools, Spark UDFs
- [ ] `area/server-infra`: MLflow Tracking server backend
- [ ] `area/tracking`: Tracking Service, tracking client APIs, autologging

Interface
- [ ] `area/uiux`: Front-end, user experience, plotting, JavaScript, JavaScript dev server
- [ ] `area/docker`: Docker use across MLflow's components, such as MLflow Projects and MLflow Models
- [ ] `area/sqlalchemy`: Use of SQLAlchemy in the Tracking Service or Model Registry
- [ ] `area/windows`: Windows support

Language
- [ ] `language/r`: R APIs and clients
- [ ] `language/java`: Java APIs and clients
- [ ] `language/new`: Proposals for new client languages

Integrations
- [ ] `integrations/azure`: Azure and Azure ML integrations
- [ ] `integrations/sagemaker`: SageMaker integrations
- [ ] `integrations/databricks`: Databricks integrations

<!--
Insert an empty named anchor here to allow jumping to this section with a fragment URL
(e.g. https://github.com/mlflow/mlflow/pull/123#user-content-release-note-category).
Note that GitHub prefixes anchor names in markdown with "user-content-".
-->
<a name="release-note-category"></a>
### How should the PR be classified in the release notes? Choose one:

- [ ] `rn/breaking-change` - The PR will be mentioned in the "Breaking Changes" section
- [ ] `rn/none` - No description will be included. The PR will be mentioned only by the PR number in the "Small Bugfixes and Documentation Updates" section
- [x] `rn/feature` - A new user-facing feature worth mentioning in the release notes
- [ ] `rn/bug-fix` - A user-facing bug fix worth mentioning in the release notes
- [ ] `rn/documentation` - A user-facing documentation change worth mentioning in the release notes

jinzhang21 commented on mlflow/pipelines/steps/train.py:
```diff
@@ -381,35 +410,71 @@ def render_schema(inputs, title):
+        run_tags = MlflowClient().get_run(run_id).data.tags
+        databricks_workspace_url = run_tags.get(MLFLOW_DATABRICKS_WORKSPACE_URL)
+        databricks_workspace_id = run_tags.get(MLFLOW_DATABRICKS_WORKSPACE_ID)
+        if databricks_workspace_url and databricks_workspace_id:
```
nit: consider `return None, None` early here if one of the condition is not True. This will makes the large chunk of code below less indented:)

jinzhang21 reviewed (approved):
Very nice! Do we plan to add a URL link to model registry on the register step card if run on Databricks as well?

dbczumar commented:
> Very nice! Do we plan to add a URL link to model registry on the register step card if run on Databricks as well?

Great idea - will add it to this PR.

sunishsheth2009 reviewed (approved):
+1 on the register card to render the link to the registered model

dbczumar commented on mlflow/pipelines/steps/evaluate.py:
```diff
@@ -6,9 +6,10 @@
 import mlflow
-from mlflow.tracking.fluent import _get_experiment_id, _set_experiment_primary_metric
-from mlflow.protos.databricks_pb2 import INVALID_PARAMETER_VALUE
+from mlflow.exceptions import MlflowException
+from mlflow.pipelines.cards import BaseCard
 from mlflow.pipelines.step import BaseStep
+from mlflow.pipelines.steps.train import TrainStep
```
Just reordered some imports

dbczumar replied:
Just reordered some imports

dbczumar commented on mlflow/pipelines/steps/evaluate.py:
```diff
@@ -24,7 +25,9 @@
 )
 from mlflow.projects.utils import get_databricks_env_vars
-from mlflow.exceptions import MlflowException
+from mlflow.protos.databricks_pb2 import INVALID_PARAMETER_VALUE
+from mlflow.tracking.fluent import _get_experiment_id, _set_experiment_primary_metric
+from mlflow.utils.databricks_utils import get_databricks_run_url
```
Just reordered some imports & added import for `get_databricks_url`

dbczumar commented on mlflow/pipelines/steps/evaluate.py:
```diff
@@ -124,7 +127,7 @@ def _run(self, output_directory):
             pipeline_root_path=self.pipeline_root,
             step_name="train",
-            relative_path="model",
+            relative_path=TrainStep.MODEL_ARTIFACT_RELATIVE_PATH,
```
Model artifact relative path needs to be referenced by other steps for model URL construction

dbczumar commented on mlflow/projects/utils.py:
```diff
@@ -362,4 +362,9 @@ def get_databricks_env_vars(tracking_uri):
     if config.ignore_tls_verification:
         env_vars["DATABRICKS_INSECURE"] = str(config.ignore_tls_verification)
+
+    workspace_info = databricks_utils.get_databricks_workspace_info_from_uri(tracking_uri)
+    if workspace_info is not None:
+        env_vars.update(workspace_info.to_environment())
```
Introduce two new internal env vars (`_DATABRICKS_WORKSPACE_HOST` and `_DATABRICKS_WORKSPACE_ID`) and set them in the subprocess used to run the train / evaluate / register steps because subprocesses cannot read this information from the command context on Databricks (command context isn't available to subprocesses).

dbczumar commented on mlflow/utils/uri.py:
```diff
@@ -252,19 +252,6 @@ def is_databricks_model_registry_artifacts_uri(artifact_uri):
     return artifact_uri_path.startswith(_MODEL_REGISTRY_ARTIFACT_URI)
 
 
-def construct_run_url(hostname, experiment_id, run_id, workspace_id=None):
```
Moved to `databricks_utils` since it's databricks-specific.

dbczumar committed:
One commit

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Submodule revert

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar commented:
@sunishsheth2009 @jinzhang21 I've addressed your comments and fixed the implementation. Links are now shown on the train, evaluate, and register step cards. Can you take another look?

dbczumar committed:
Fix client tests

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
fix ci

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Remove unused / invalid import

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Put run summary back to end, fix tests

Signed-off-by: dbczumar <corey.zumar@databricks.com>

dbczumar committed:
Use mocks instead

Signed-off-by: dbczumar <corey.zumar@databricks.com>
